File: src/api/reencuentro_api/services/ciudades.py
"""Zonas cubiertas por Reencuentro: fuente de verdad de los bounding boxes.

Función pura / datos estáticos (sin I/O). Coordenadas reales aproximadas del
casco urbano de cada ciudad. `COLOMBIA` es el bounding box continental del país
(sin San Andrés): se usa como vista agregada del mapa y como lienzo para
reportar desde cualquier lugar del país (zona "Otro").

Duplicada conscientemente en `src/web/src/lib/ciudades.ts` (el frontend no
consulta estos valores por HTTP) — si cambian aquí, cambiar allá también.
"""

ZONAS: dict[str, dict[str, float]] = {
    "Armenia": {
        "lat_min": 4.49,
        "lat_max": 4.58,
        "lng_min": -75.75,
        "lng_max": -75.63,
        "centro_lat": 4.534,
        "centro_lng": -75.681,
    },
    "Pereira": {
        "lat_min": 4.77,
        "lat_max": 4.85,
        "lng_min": -75.76,
        "lng_max": -75.65,
        "centro_lat": 4.813,
        "centro_lng": -75.696,
    },
    "Manizales": {
        "lat_min": 5.02,
        "lat_max": 5.12,
        "lng_min": -75.55,
        "lng_max": -75.44,
        "centro_lat": 5.070,
        "centro_lng": -75.514,
    },
    "Cali": {
        "lat_min": 3.33,
        "lat_max": 3.50,
        "lng_min": -76.60,
        "lng_max": -76.44,
        "centro_lat": 3.452,
        "centro_lng": -76.532,
    },
    "Quibdó": {
        "lat_min": 5.65,
        "lat_max": 5.72,
        "lng_min": -76.68,
        "lng_max": -76.62,
        "centro_lat": 5.695,
        "centro_lng": -76.661,
    },
    "Bogotá": {
        "lat_min": 4.55,
        "lat_max": 4.80,
        "lng_min": -74.20,
        "lng_max": -74.00,
        "centro_lat": 4.65,
        "centro_lng": -74.08,
    },
}

COLOMBIA: dict[str, float] = {
    "lat_min": -4.3,
    "lat_max": 12.6,
    "lng_min": -79.1,
    "lng_max": -66.9,
    "centro_lat": 4.6,
    "centro_lng": -74.1,
}

# Reportes desde fuera de las zonas con bounding box propio: el pin se pone
# sobre el mapa nacional y la ciudad real va en texto libre (Report.ciudad_texto).
ZONA_OTRO = "Otro"


def zona_valida(zona: str) -> bool:
    return zona in ZONAS or zona == ZONA_OTRO

File: src/web/src/lib/ciudades.ts
// Zonas cubiertas: bounding box + centro de cada una, más el lienzo nacional.
// Duplicado consciente de src/api/reencuentro_api/services/ciudades.py (el
// frontend no consulta estos valores por HTTP) — si cambian allá, cambiar aquí.

export type BoundingBox = {
  latMin: number;
  latMax: number;
  lngMin: number;
  lngMax: number;
  centroLat: number;
  centroLng: number;
};

export const ZONAS: Record<string, BoundingBox> = {
  Armenia: {
    latMin: 4.49,
    latMax: 4.58,
    lngMin: -75.75,
    lngMax: -75.63,
    centroLat: 4.534,
    centroLng: -75.681,
  },
  Pereira: {
    latMin: 4.77,
    latMax: 4.85,
    lngMin: -75.76,
    lngMax: -75.65,
    centroLat: 4.813,
    centroLng: -75.696,
  },
  Manizales: {
    latMin: 5.02,
    latMax: 5.12,
    lngMin: -75.55,
    lngMax: -75.44,
    centroLat: 5.07,
    centroLng: -75.514,
  },
  Cali: {
    latMin: 3.33,
    latMax: 3.5,
    lngMin: -76.6,
    lngMax: -76.44,
    centroLat: 3.452,
    centroLng: -76.532,
  },
  Quibdó: {
    latMin: 5.65,
    latMax: 5.72,
    lngMin: -76.68,
    lngMax: -76.62,
    centroLat: 5.695,
    centroLng: -76.661,
  },
  Bogotá: {
    latMin: 4.55,
    latMax: 4.8,
    lngMin: -74.2,
    lngMax: -74.0,
    centroLat: 4.65,
    centroLng: -74.08,
  },
  // Valle de Aburrá completo: Medellín + Bello (norte) e Itagüí/Envigado (sur).
  Medellín: {
    latMin: 6.13,
    latMax: 6.36,
    lngMin: -75.66,
    lngMax: -75.5,
    centroLat: 6.244,
    centroLng: -75.581,
  },
};

// Lienzo nacional (continental, sin San Andrés): vista agregada del mapa y
// fallback para reportar desde cualquier lugar del país (zona "Otro").
export const COLOMBIA: BoundingBox = {
  latMin: -4.3,
  latMax: 12.6,
  lngMin: -79.1,
  lngMax: -66.9,
  centroLat: 4.6,
  centroLng: -74.1,
};

export const ZONA_OTRO = 'Otro';
export const NOMBRES_ZONAS = Object.keys(ZONAS);

// Landings por zona (feature 46): slug de URL → nombre de zona. Espejo del
// SLUG_ZONAS del backend (routers/paginas.py).
export const SLUGS_ZONA: Record<string, string> = {
  cali: 'Cali',
  armenia: 'Armenia',
  pereira: 'Pereira',
  manizales: 'Manizales',
  quibdo: 'Quibdó',
  bogota: 'Bogotá',
  medellin: 'Medellín',
};

// Ciudades de Colombia para el registro (feature 17): las zonas con mapa propio
// van primero en el select; este resto cubre todas las capitales departamentales
// más las ciudades grandes que no son capital, en orden alfabético.
export const OTRAS_CIUDADES_COLOMBIA = [
  'Arauca',
  'Barrancabermeja',
  'Barranquilla',
  'Bello',
  'Bucaramanga',
  'Buenaventura',
  'Cartagena',
  'Cúcuta',
  'Dosquebradas',
  'Florencia',
  'Ibagué',
  'Inírida',
  'Leticia',
  'Mitú',
  'Mocoa',
  'Montería',
  'Neiva',
  'Palmira',
  'Pasto',
  'Popayán',
  'Puerto Carreño',
  'Riohacha',
  'San Andrés',
  'San José del Guaviare',
  'Santa Marta',
  'Sincelejo',
  'Soacha',
  'Soledad',
  'Tuluá',
  'Tunja',
  'Valledupar',
  'Villavicencio',
  'Yopal',
];

// El lienzo a usar para una zona dada: las 6 conocidas usan su propio bounding
// box; "Otro" (y la vista "Colombia" del mapa) usan el nacional.
export function cajaDeZona(zona: string): BoundingBox {
  return ZONAS[zona] ?? COLOMBIA;
}

// Geolocalización (feature 31): ¿dónde caen unas coords reales?
function dentroDeCaja(caja: BoundingBox, lat: number, lng: number): boolean {
  return lat >= caja.latMin && lat <= caja.latMax && lng >= caja.lngMin && lng <= caja.lngMax;
}

export function coordsEnZona(zona: string, lat: number, lng: number): boolean {
  return dentroDeCaja(cajaDeZona(zona), lat, lng);
}

export function zonaQueContiene(lat: number, lng: number): string | null {
  for (const [nombre, caja] of Object.entries(ZONAS)) {
    if (dentroDeCaja(caja, lat, lng)) return nombre;
  }
  return null;
}
